processor/spanprocessor/span_processor_test.go

```diff
@@ -0,0 +1,620 @@
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector/component"
+	"github.com/open-telemetry/opentelemetry-collector/component/componenterr"
+	"github.com/open-telemetry/opentelemetry-collector/exporter/exportertest"
+	"github.com/open-telemetry/opentelemetry-collector/internal/data"
+	"github.com/open-telemetry/opentelemetry-collector/internal/data/testdata"
+	"github.com/open-telemetry/opentelemetry-collector/internal/processor/filterspan"
+	"github.com/open-telemetry/opentelemetry-collector/translator/conventions"
+)
+
+func TestNewTraceProcessor(t *testing.T) {
+	factory := Factory{}
+	cfg := factory.CreateDefaultConfig()
+	oCfg := cfg.(*Config)
+	tp, err := newSpanProcessor(nil, *oCfg)
+	require.Error(t, componenterr.ErrNilNextConsumer, err)
+	require.Nil(t, tp)
+
+	tp, err = newSpanProcessor(exportertest.NewNopTraceExporter(), *oCfg)
+	require.Nil(t, err)
+	require.NotNil(t, tp)
+}
+
+// Common structure for the test cases.
+type testCase struct {
+	serviceName      string
+	inputName        string
+	inputAttributes  map[string]data.AttributeValue
+	outputName       string
+	outputAttributes map[string]data.AttributeValue
+}
+
+// runIndividualTestCase is the common logic of passing trace data through a configured attributes processor.
+func runIndividualTestCase(t *testing.T, tt testCase, tp component.TraceProcessor) {
+	t.Run(tt.inputName, func(t *testing.T) {
+		td := generateTraceData(tt.serviceName, tt.inputName, tt.inputAttributes)
+
+		assert.NoError(t, tp.ConsumeTrace(context.Background(), td))
+		// Ensure that the modified `td` has the attributes sorted:
+		rss := td.ResourceSpans()
+		for i := 0; i < rss.Len(); i++ {
+			rs := rss.At(i)
+			if rs.IsNil() {
+				continue
+			}
+			if !rs.Resource().IsNil() {
+				rs.Resource().Attributes().Sort()
+			}
+			ilss := rss.At(i).InstrumentationLibrarySpans()
+			for j := 0; j < ilss.Len(); j++ {
+				ils := ilss.At(j)
+				if ils.IsNil() {
+					continue
+				}
+				spans := ils.Spans()
+				for k := 0; k < spans.Len(); k++ {
+					s := spans.At(k)
+					if !s.IsNil() {
+						s.Attributes().Sort()
+					}
+				}
+			}
+		}
+		assert.EqualValues(t, generateTraceData(tt.serviceName, tt.outputName, tt.outputAttributes), td)
+	})
+}
+
+func generateTraceData(serviceName, inputName string, attrs map[string]data.AttributeValue) data.TraceData {
+	td := data.NewTraceData()
+	td.ResourceSpans().Resize(1)
+	rs := td.ResourceSpans().At(0)
+	if serviceName != "" {
+		rs.Resource().InitEmpty()
+		rs.Resource().Attributes().Upsert(
+			data.NewAttributeKeyValueString(conventions.AttributeServiceName, serviceName))
+	}
+	rs.InstrumentationLibrarySpans().Resize(1)
+	ils := rs.InstrumentationLibrarySpans().At(0)
+	spans := ils.Spans()
+	spans.Resize(1)
+	spans.At(0).SetName(inputName)
+	spans.At(0).Attributes().InitFromMap(attrs).Sort()
+	return td
+}
+
+// TestSpanProcessor_Values tests all possible value types.
+func TestSpanProcessor_NilEmptyData(t *testing.T) {
```

Comment:
TODO: Add tests for:
* one-empty-one-nil-resource-span
* one-empty-one-nil-instrumentation-library
* one-span-one-nil.